Content.Server/EntityEffects/Effects/ActivateRandom.cs

```diff
@@ -0,0 +1,166 @@
+    {
+        return Loc.GetString("reagent-effect-guidebook-activate-random",
+            ("chance", Probability),
+            ("cases", string.Join("\n", Cases.ToList().Select(case_ =>
+                "- "
+                + string.Join(", ", case_.Activates.ToList()
+                    .Select(activate =>
+                        prototype.TryIndex(activate.Reagent.Prototype, out ReagentPrototype? reagentProto)
+                            ? Loc.GetString("reagent-effect-guidebook-activate-random-activate",
+                                ("factor", activate.Factor),
+                                ("catalyst", !activate.Consume),
+                                ("reagent", reagentProto.LocalizedName))
+                            : null)
+                        .Where(x => x is not null))
+                + ":\n"
+                + string.Join("\n", case_.Effects.ToList()
+                    .Select(effect =>
+                    {
+                        var desc = effect.GuidebookEffectDescription(prototype, entSys);
+                        return desc is null ? null : "  " + desc;
+                    })
+                    .Where(x => x is not null))))
+                + "\n"));
+    }
+
+    public override void Effect(EntityEffectBaseArgs baseArgs)
+    {
+         // ActivateRandom works by activating other reagents, it needs access to extended information
+        if (baseArgs is not EntityEffectReagentArgs)
+            return;
+        EntityEffectReagentArgs args = (EntityEffectReagentArgs) baseArgs;
+            
+        if (args.Source is null)
+            return; // ActivateRandom activates chemicals from the args.Source.
+            // If source is not present, ActivateRandom has nothing to activate.
+
+        // Only one random set of chemicals is activated. The chance to activate
+        // some speicific set is based on how many of these chemicals (with respect to factor)
+        // are present in the solution.
+        // We use chance and chanceSum to calculate proprotions of chemical sets (with respect to factor)
+        // present in the solution.
+        float[] chances = new float[Cases.Length];
+        float chancesSum = 0f;
+
+        for (int i = 0; i < Cases.Length; i++)
+        {
+            var _case = Cases[i];
+            foreach (var activate in _case.Activates)
+            {
+                if (args.Source.TryGetReagent(activate.Reagent, out var quantity))
+                {
+                    var impact = quantity.Quantity.Float() * activate.Factor.Float();
+                    chances[i] += impact;
+                    chancesSum += impact;
+                }
+            }
+        }
+
+        // Next, select a "random" set. Sets of higher propotions are prioritized.
+        int index = 0;
+        IRobustRandom _random = IoCManager.Resolve<IRobustRandom>();
+        {
```

Comment:
You should avoid doing IoC resolves as they use thread-locals, especially in hot loops. Just put it as a dependency on the effect.

Reply:
How can this be done? I tried using 
```csharp
[Dependency] private readonly IRobustRandom _random = default!;
```
before, but it doesn't seem to work on `EntityEffect`s.